File: python/Figure3_1.py
# Author: Kenji Kashima
# Date  : 2025/04/01

import numpy as np
import matplotlib.pyplot as plt
import config

np.random.seed(100)

def figure3_1a(k_bar=9):
    '''
        k_bar - number of time steps
    '''
    figsize = config.global_config(type=1)

    n_tmp = 1000
    n_x = 2 * n_tmp + 1   # number of discrete x points over [-1, 1]
    x = np.linspace(-1, 1, n_x)
    dx = x[1] - x[0]      # discretization interval; i'th cell: [ x[i]-dx/2, x[i]+dx/2 ]

    # Build transition probability matrix P (approximate uniform noise over interval)
    P = np.zeros([n_x, n_x])
    for i in range(n_x):
        # fb(x) = x + 0.1*(x - x^3)  
        # noise amplitude: c(x) = 0.5*(1 - |x|)  
        upper = x[i] + 0.1 * (x[i] - x[i]**3) + 0.5 * (1 - np.abs(x[i]))
        lower = x[i] + 0.1 * (x[i] - x[i]**3) - 0.5 * (1 - np.abs(x[i]))
        idx_u = int(np.floor(upper * n_tmp + 0.5)) + n_tmp
        idx_l = int(np.ceil(lower * n_tmp - 0.5)) + n_tmp
        for j in range(idx_l, idx_u + 1):
            P[j, i] += 1.0 / (idx_u - idx_l + 1)

    # Define initial distribution φₓ₀ = 1 on [-0.5, 0.5]
    init = np.where((x >= -0.5) & (x <= 0.5), 1.0, 0.0)
    # Normalize so that sum(init)*dx = 1
    init /= (np.sum(init) * dx)

    # Initialize phi array: each column is the state distribution at time k
    phi = np.zeros([n_x, k_bar + 1])
    phi[:, 0] = init

    # Propagate the distribution by multiplying with P at each time step
    for k in range(k_bar):
        phi[:, k + 1] = P @ phi[:, k]
        # If needed, re-normalize to guard against numerical drift:
        # phi[:, k+1] /= (np.sum(phi[:, k+1]) * dx)

    # Plot from k=0 onward
    fig, ax = plt.subplots(subplot_kw={"projection": "3d"}, figsize=figsize)
    for k in range(k_bar + 1):
        ax.plot3D(np.full(n_x, k), x, phi[:, k])

    ax.set_xlabel(r'$k$', labelpad=20)
    ax.set_ylabel(r'${\rm x}$', labelpad=20)
    ax.zaxis.set_rotate_label(False)
    ax.set_zlabel(r'$\varphi_{x_k}$', rotation=90, labelpad=20)
    ax.view_init(26, -107)
    ax.set_yticks([-1, -0.5, 0, 0.5, 1])
    plt.grid()
    plt.tight_layout()
    plt.savefig("./Figure3_1a.pdf")
    plt.show()


def figure3_1b(k_bar: int = 50, n_sample: int = 10):
    '''
        k_bar       - total number of steps
        n_sample  - number of sample trajectories
    '''
    config.global_config()

    plt.figure(figsize=(8, 7))
    for _ in range(n_sample):
        x = np.zeros(k_bar)
        x[0] = np.random.rand() - 0.5  # initial state in [-0.5, 0.5]
        for k in range(k_bar - 1):
            noise = np.random.rand() - 0.5  # uniform in [-0.5, 0.5]
            x[k + 1] = x[k] + 0.1 * (x[k] - x[k]**3) + noise * (1 - np.abs(x[k]))
        plt.plot(x, linewidth=2)

    plt.xlim([0, k_bar - 1])
    plt.xlabel(r'$k$')
    plt.ylabel(r'$x_k$')
    plt.grid()
    plt.ylim([-1, 1])
    plt.tight_layout()
    plt.savefig("./Figure3_1b.pdf")
    plt.show()


if __name__ == '__main__':
    figure3_1a(k_bar=9)
    figure3_1b(k_bar=30, n_sample=10)

File: matlab/Figure3_1.m
% Author: Kenji Kashima
% Date  : 2025/04/01

clear; close all;
rng(1);  % keep original MATLAB seed

set(groot, 'defaultAxesTickLabelInterpreter','latex');
set(groot, 'defaultTextInterpreter','latex');
set(groot, 'defaultLegendInterpreter','latex');

%% Parameters
n_tmp    = 1000;
n_x      = 2 * n_tmp + 1;   % number of x grid points
x        = linspace(-1, 1, n_x)';
dx       = x(2) - x(1);
n_k_a    = 9;               % number of time steps for Figure 3.1(a)
n_k_b    = 50;              % total steps for Figure 3.1(b)
n_sample = 10;              % number of sample trajectories

% Build transition probability matrix P (approximate uniform noise over interval)
P = zeros(n_x, n_x);
for i = 1:n_x
    xi    = x(i);
    fb    = xi + 0.1*(xi - xi^3);          % fb(x)
    noise = 0.5*(1 - abs(xi));             % c(x)
    upper = fb + noise;
    lower = fb - noise;
    % Python: idx_u = int(floor(upper*n_tmp + 0.5)) + n_tmp
    idx_u = floor(upper * n_tmp + 0.5) + n_tmp + 1;
    % Python: idx_l = int(ceil(lower*n_tmp - 0.5)) + n_tmp
    idx_l = ceil(lower * n_tmp - 0.5) + n_tmp + 1;
    for j = idx_l:(idx_u)
        P(j,i) = P(j,i) + 1.0 / (idx_u - idx_l + 1);
    end
end

%% Figure 3.1(a): propagate a uniform init on [-0.5,0.5]
% define and normalize initial distribution
init = double((x >= -0.5) & (x <= 0.5));
init = init / (sum(init) * dx);

% propagate
phi = zeros(n_x, n_k_a+1);
phi(:,1) = init;
for k = 1:n_k_a
    phi(:,k+1) = P * phi(:,k);
end

% plot
figure('Name','Figure 3.1(a)'); hold on; view(26, -107); grid on;
for k = 0:n_k_a
    plot3(...
      k*ones(n_x,1), ...
      x, ...
      phi(:,k+1), ...
      'LineWidth', 2 ...
    );
end
xlabel('$k$','Interpreter','latex','FontSize',18);
ylabel('$x$','Interpreter','latex','FontSize',18);
zlabel('$\varphi_{x_k}$','Interpreter','latex','FontSize',18);
yticks([-1 -0.5 0 0.5 1]);
movegui('northeast');

%% Figure 3.1(b): sample trajectories with matching noise scale
figure('Name','Figure 3.1(b)'); hold on; grid on;
for s = 1:n_sample
    xx = zeros(1, n_k_b);
    xx(1) = rand - 0.5;  % in [-0.5, 0.5]
    for k = 1:(n_k_b-1)
        noise = rand - 0.5;  % uniform in [-0.5,0.5]
        xi    = xx(k);
        xx(k+1) = xi + 0.1*(xi - xi^3) + noise*(1 - abs(xi));
    end
    plot(0:(n_k_b-1), xx, 'LineWidth', 2);
end
xlim([0, n_k_b-1]);
xlabel('$k$','Interpreter','latex','FontSize',18);
ylabel('$x_k$','Interpreter','latex','FontSize',18);
movegui('northwest');
